lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -0,0 +1,415 @@
+	pub(crate) fn from_slice(s: &[u8]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[u8] {
+		if self.datalen == 0 { return &[]; }
+		unsafe { std::slice::from_raw_parts(self.data, self.datalen) }
+	}
+}
+
+#[repr(C)]
+pub struct usizeslice {
+	pub data: *const usize,
+	pub datalen: usize
+}
+impl usizeslice {
+	pub(crate) fn from_slice(s: &[usize]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[usize] {
+		if self.datalen == 0 { return &[]; }
+		unsafe { std::slice::from_raw_parts(self.data, self.datalen) }
+	}
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// Arbitrary 32 bytes, which could represent one of a few different things. You probably want to
+/// look up the corresponding function in rust-lightning's docs.
+pub struct ThirtyTwoBytes {
+	pub data: [u8; 32],
+}
+impl ThirtyTwoBytes {
+	pub(crate) fn empty() -> Self {
+		Self { data: [0; 32] }
+	}
+}
+
+#[repr(C)]
+pub struct ThreeBytes { pub data: [u8; 3], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct FourBytes { pub data: [u8; 4], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct TenBytes { pub data: [u8; 10], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct SixteenBytes { pub data: [u8; 16], }
+
+pub(crate) struct VecWriter(pub Vec<u8>);
+impl lightning::util::ser::Writer for VecWriter {
+	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
+		self.0.extend_from_slice(buf);
+		Ok(())
+	}
+	fn size_hint(&mut self, size: usize) {
+		self.0.reserve_exact(size);
+	}
+}
+pub(crate) fn serialize_obj<I: lightning::util::ser::Writeable>(i: &I) -> derived::CVec_u8Z {
+	let mut out = VecWriter(Vec::new());
+	i.write(&mut out).unwrap();
+	CVecTempl::from(out.0)
+}
+pub(crate) fn deserialize_obj<I: lightning::util::ser::Readable>(s: u8slice) -> Result<I, lightning::ln::msgs::DecodeError> {
+	I::read(&mut s.to_slice())
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// A Rust str object, ie a reference to a UTF8-valid string.
+/// This is *not* null-terminated so cannot be used directly as a C string!
+pub struct Str {
+	pub chars: *const u8,
+	pub len: usize
+}
+impl Into<Str> for &'static str {
+	fn into(self) -> Str {
+		Str { chars: self.as_ptr(), len: self.len() }
+	}
+}
+impl Into<&'static str> for Str {
+	fn into(self) -> &'static str {
+		if self.len == 0 { return ""; }
+		std::str::from_utf8(unsafe { std::slice::from_raw_parts(self.chars, self.len) }).unwrap()
+	}
+}
+
+// Note that the C++ headers memset(0) all the Templ types to avoid deallocation!
+// Thus, they must gracefully handle being completely null in _free.
+
+/// CSliceTempl may have the word slice in it, but its really a Vec. Because its used for things
+/// that require per-item conversion, we have to use a new buffer for it. Of course for opaque
+/// types, each item will simply be a pointer to the original slice's corresponding item, so we may
+/// not own the actual Rust item, only the buffer.
```

Comment:
"so we may not own the actual Rust item, only the buffer."
What does that mean?
I personally think it should be renamed to COwnedVecTemp<T> or something like that to show that it is owned and if zeroed out then it leaks the buffer

Reply:
Is the below more clear? In general, we map Rust structs and other things which we want to be opaque as `CStruct { inner: *mut RustStruct, ...}`. Thus, whats meant here is that the Rust objects may not be owned, just pointers to them. Of course in doing so we make sure the pointers don't get free'd, but that's captured in the struct mapping.
```
/// CSliceTempl may have the word slice in it, but its really a Vec. Because its used for things
/// that require per-item conversion, we have to use fresh memory for it. Of course for opaque
/// types, each item will simply be a pointer to the original slice's corresponding item, so we may
/// not own the actual Rust object, only the memory pointed to by data.
```

Reply:
Yeah I think so